packages/tools/src/tools/segmentation/PaintFillTool.ts

```diff
@@ -77,18 +85,64 @@ class PaintFillTool extends BaseTool {
     }
 
     const { segmentationId, type } = activeSegmentationRepresentation;
+    if (type === SegmentationRepresentations.Contour) {
+      throw new Error('Not implemented yet');
+    }
+
     const segmentIndex =
       segmentIndexController.getActiveSegmentIndex(segmentationId);
     const segmentsLocked: number[] =
       segmentLocking.getLockedSegments(segmentationId);
     const { representationData } = getSegmentation(segmentationId);
 
-    const { volumeId } = representationData[type] as LabelmapSegmentationData;
-    const segmentation = cache.getVolume(volumeId);
-    const { dimensions, direction } = segmentation;
-    const scalarData = segmentation.getScalarData();
+    const labelmapData =
+      representationData[SegmentationRepresentations.Labelmap];
+
+    let dimensions, direction, scalarData, imageData;
+    const segmentationType = labelmapData.type;
+    const auxiliaryData = { type: '', imageIds: [] };
+    if (segmentationType === 'volume' || !segmentationType) {
+      const { volumeId } = representationData[
+        type
+      ] as LabelmapSegmentationDataVolume;
+      const segmentation = cache.getVolume(volumeId);
+      dimensions = segmentation.dimensions;
+      direction = segmentation.direction;
+      scalarData = segmentation.getScalarData();
+      imageData = segmentation.imageData;
+    } else {
+      const { referencedImageIds } =
+        labelmapData as LabelmapSegmentationDataStack;
+      const {
+        zSpacing,
+        sortedImageIds,
+        origin,
+        direction: calculatedDirection,
+      } = sortImageIds(referencedImageIds);
+      const currentImageId = viewport.getCurrentImageId();
+
+      imageData = createVTKImageDataFromImageId(currentImageId);
```

Comment:
This is not consistent with what is happening in our volumes, you are here actually creating the imageData though in volumes the segmentation should be pre-created and above we only grab it by `cache.getVolume`. I think it is fine but we need to document this much better, also the following couple of lines are shared between all tools so THIS sure should get refactored